agent/cache/cache.go

```diff
@@ -609,13 +604,91 @@ func (c *Cache) fetch(key string, r getOptions, allowNew bool, attempt uint, ign
 	// identical calls to fetch will return the same waiter rather than
 	// perform multiple fetches.
 	entry.Fetching = true
+	entry.HasGoroutine = true
 	c.entries[key] = entry
 	metrics.SetGauge([]string{"consul", "cache", "entries_count"}, float32(len(c.entries)))
 	metrics.SetGauge([]string{"cache", "entries_count"}, float32(len(c.entries)))
 
-	tEntry := r.TypeEntry
 	// The actual Fetch must be performed in a goroutine.
-	go func() {
+	go c.launchBackgroundFetcher(key, r, entry)
+
+	return entry.Waiter
+}
+
+func (c *Cache) launchBackgroundFetcher(key string, r getOptions, initialEntry cacheEntry) {
+	defer c.mutateExistingEntry(key, r, func(entry *cacheEntry) {
+		if entry.HasGoroutine {
+			entry.HasGoroutine = false
+			entry.Fetching = false
+		}
+	})
+
+	var (
+		attempt uint
+		entry   = initialEntry
+	)
+	for {
+		if !c.runBackgroundFetcherOnce(key, r, entry, &attempt) {
+			return
+		}
+
+		var totalWait time.Duration
+
+		// If we're over the attempt minimum, start an exponential backoff.
+		if wait := backOffWait(attempt); wait > 0 {
+			totalWait += wait
+		}
+
+		// If we have a timer, wait for it
+		if r.TypeEntry.Opts.RefreshTimer > 0 {
+			totalWait += r.TypeEntry.Opts.RefreshTimer
+		}
```

Comment:
By my comment on the former PR, those 2 could be simplified by removing the if statements.
```suggestion

		// If we're over the attempt minimum, start an exponential backoff.
		totalWait := backOffWait(attempt)

		// If we have a timer, wait for it
		totalWait += r.TypeEntry.Opts.RefreshTimer
		
```